### PR TITLE
[8.x] Improved Maintenance Mode

### PR DESCRIPTION
**Pull request by taylorotwell**
This PR implements a much improved maintenance mode.

### Prerendered Templates

One main deficiency of the previous implementation was the entire framework has to be booted in order to render the maintenance page. This could cause various problems if Composer was in the middle of updating dependencies, an infrastructure component was down, etc.

The new `render` option of the `down` command allows you to prerender a view. In addition, a new `bootstrap/maintenance.php` file is written. This file is loaded before Composer in the framework's `index.php` file and is able to render the template and exit with zero dependencies. Composer and Laravel are never loaded.

```
php artisan down --render="errors::503"
```

### Secret Bypass Instead Of IP Allow List

The IP allow list has been removed in favor of a secret bypass route. The IP white list isn't practical for many applications. You may now use the `secret` option of the `down` command:

```
php artisan down --secret=MvYgEHudaikRaX4JRcyh
```

Once the application is in maintenance mode, you may navigate to the secret's value as a route. For example:

https://example.com/MvYgEHudaikRaX4JRcyh

The maintenance middleware will intercept this request and issue a maintenance mode bypass cookie to the user and redirect them to `/`, allowing them to access the application and bypass maintenance mode for 12 hours. The cookie contains a HMAC signed expiration time.

### Redirects

A new `redirect` option has been added to the `down` command to instruct all paths of the application to redirect to the given path during maintenance:

```
php artisan down --redirect=/
```

### Status Code

A new `status` option has been added to the `down` command to allow the status code to be customized. The default is 503. You can combine this with the `redirect` and `secret` options to still show a normal marketing / "coming soon" site but allow the real application to be accessed via the secret...

```
php artisan down --redirect=/ --status=200 --secret=MvYgEHudaikRaX4JRcyh
```

### CODE REVIEW

**Comment by lstables** [comment]
That is awesome Mr @taylorotwell always wanted this kinda thing instead of doing “private” routes.

**Comment by mattstauffer** [comment]
I love this so much :)

**Comment by stevebauman** [comment]
This is awesome 🙌

**Comment by fgilio** [comment]
This is fantastic, thanks!

**Comment by ademakanaky** [comment]
This is a good one 👍

**Comment by georgeboot** [comment]
How will this work in environments like Vapor as we can't persist a file to the file system?

File: src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php [hunk 1]
```diff
@@ -49,20 +49,60 @@ public function handle($request, Closure $next)
...
         }
 
         return $next($request);
     }
 
+    /**
+     * Determine if the incoming request has a maintenance mode bypass cookie.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return bool
```
Comment (voyula) [inline src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:94]: Add @param  array  $data

File: src/Illuminate/Foundation/Console/DownCommand.php [hunk 2]
```diff
@@ -64,13 +69,40 @@ public function handle()
...
     }
 
+    /**
+     * Get the path that users should be redirected to.
+     *
+     * @return string
+     */
+    protected function redirectPath()
+    {
+        if ($this->option('redirect') && $this->option('redirect') !== '/') {
```
Comment (mohammed-adil-moughal) [inline src/Illuminate/Foundation/Console/DownCommand.php:87]: Note - looks good but generally not a fan of directly comparing strings, would prefer str cmp methods instead

Comment (mfn) [reply src/Illuminate/Foundation/Console/DownCommand.php:87]: Why?

FTR, there's exactly _one_ `strcmp` in the entire framework AFAICS. I didn't bother to count the direct string comparison checks 😄

**Comment by jesseleite** [comment]
I can get `down` with this!

**Pull request merged**